overlord/servicestate/quota_control.go

```diff
@@ -68,8 +71,82 @@ func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[s
 		ensureOpts.RequireMountedSnapdSnap = true
 	}
 
-	// TODO: do we need to restart modified services ?
-	return wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, nil, progress.Null)
+	grpsToRestart := []*quota.Group{}
+	appsToRestartBySnap := map[*snap.Info][]*snap.AppInfo{}
+
+	collectModifiedUnits := func(app *snap.AppInfo, grp *quota.Group, unitType string, name, old, new string) {
+		switch unitType {
+		case "slice":
+			// this slice was modified and needs to be restarted
+			grpsToRestart = append(grpsToRestart, grp)
+
+		case "service":
+			sn := app.Snap
+			appsToRestartBySnap[sn] = append(appsToRestartBySnap[sn], app)
+
+			// TODO: what about sockets and timers? activation units just start
+			// the full unit, so as long as the full unit is restarted we should
+			// be okay?
+		}
+	}
+	if err := wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, collectModifiedUnits, progress.Null); err != nil {
+		return err
+	}
+
+	// now restart the slices
```

Comment:
Should we check if we are preseeding here and return? Or it's not going to be ever triggered during preseeding?